### ISSUE TITLE
Object GROUP templates

### ISSUE THREAD

**Issue by witlock** [bjorn/tiled#1860]
When I first heard about templates on Patreon I was hoping we could create templates of object layers/groups, but it seems like it's just for objects. For example a tree in my game is created from 5 - 6 objects and each tree has it's own layer (colliders, trunk image, top image etc.). Right now I'm duplicating the layers and it works, but loading them from templates would be pretty nice. Is this something we could see in the future or is this something only I am doing?

**Comment by bjorn** [comment]
While I can see the use-case, I think the solution is going in the wrong direction. This is because as it is now, it will only add complexity. I'd much rather re-think the data structures such that we can simplify things while adding the needed functionality for compound objects, like your trees.

Of course it's not unusual for an object to consist of several parts. In most game engines today, we don't talk about objects but about entities. And entities can have multiple components. Your tree entity could have several image components and a collider component. This is the direction I'd prefer to go into, but it could become quite a big change.

**Comment by phil123456** [comment]
I agree, templates imply talking about group of objects, no matter you call them teplates or entities

actually if it was my editor project, I'd clearly not hesitate and I would also make possible the entity having parts drawned on different layers, it would be a killer feature :-)

and indeed it would be a group of objects that end up creating an entity, it would also define the physical (collision) properties the entity... 

ex:
- a tree has 2 parts (ground and on top of everything layer) + collision circle for instance

- I am writing a game mockup that has maps like in ultima VIII (but not isometric, just top down) and oh yeah, beeing able to draw a whole house with like 2 or 3 levels.. (even have the dimensions of the the houses beeing parameters and drawn on the map at first) 

of course it depends on your game needs but killer features make great interesting rich possibilities

**Comment by bjorn** [comment]
> I agree, templates imply talking about group of objects, no matter you call them templates or entities

I think first you need to support compound objects, and then you can support templates of them. As it is, if objects could be compound, the template system will almost automatically support that as well. Supporting templates that instantiate groups of objects is something else entirely, though also interesting to consider.

> actually if it was my editor project, I'd clearly not hesitate and I would also make possible the entity having parts drawned on different layers, it would be a killer feature :-)

Tiled being free software, please don't think you need to convince me about such a change first. I'm hesitant because I think it needs careful consideration and because it will be a major effort to get it done in a clean way. But Tiled is yours to change around as you see fit and I'd love to see what you'll do with it.

**Comment by X-Ray-Jin** [comment]
I think this really would be a big thing.
Speaking of entities:
For our project I am actually kind of working on a LibGDX - Tiled entity system and an interface that extends the basic loader and makes using Tiled in LibGDX more intuitive. For now it is not standalone. However while making the game I already started splitting everything nice so my plan is at some point to separate it completely.

I am actually doing it quite similar as mentioned above with an entity class and components which at the moment we call logics for some reason. Anyways I am glad that you plan to go that direction. We are using special tilesets to define components at the moment to keep them reasonable separate and maintain a good overview. (@bjorn related to the other discussion #1930 not having just one loading thing)

I really would like to stay informed about your plans on this so I can adept to it as soon as possible.
